File: problem-solving/_lc-121-easy-best-time-to-buy-and-sell-stock/1-solution.py
"""
https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

Complexity
----------
Time: O(n)
Space: O(1)
"""

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        diff = 0
        localMin = prices[0]
        for i in range(1, len(prices)):
            if diff < prices[i] - localMin:
                diff = prices[i] - localMin
            if prices[i] < localMin:
                localMin = prices[i]
        return diff

File: problem-solving/_lc-121-easy-best-time-to-buy-and-sell-stock/1-solution.go
/*
https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

Complexity
----------
Time: O(n)
Space: O(1)
*/

func maxProfit(prices []int) int {
    buy, profit := prices[0], 0
    for i := 1; i < len(prices); i++ {
        if profit < prices[i] - buy {
            profit = prices[i] - buy
        }
        if prices[i] < buy {
            buy = prices[i]
        }
    }
    return profit
}
